content/en/docs/guides/performance.md

```diff
@@ -0,0 +1,102 @@
+    handler on the server side. Streams are particularly well-suited to handling
+    a long-lived logical flow of data from the client-to-server,
+    server-to-client, or in both directions. Streams, however, cannot be load
+    balanced once they have started and can be hard to debug for stream
+    failures. \
+    ***Side note:*** *This does not apply to Python (see Python section for
+    details).*
+
+*   Each gRPC channel uses 0 or more HTTP/2 connections and each connection
+    usually has a limit on the number of concurrent streams. When the number of
+    active RPCs on the connection reaches this limit, additional RPCs are queued
+    in the client and must wait for active RPCs to finish before they are sent.
+    Applications with high load or long-lived streaming RPCs might see
+    performance issues because of this queueing. There are two possible
+    solutions:
+
+    1.  **Create a separate channel for each area of high load** in the
+        application, or
+
+    2.  **Use a pool of gRPC channels** to randomly distribute RPCs over
+        multiple connections (channels must have different channel args to
+        prevent re-use so define a use-specific channel arg such as channel
+        number). \
+        ***Side note:*** *The gRPC team has plans to add a feature to fix these
+        performance issues, so any solution involving creating multiple channels
+        is a temporary workaround that should eventually not be needed.*
+
+### C++
+
+*   **Do not use Sync API for performance sensitive servers.** If performance
+    and/or resource consumption are not concerns, use the Sync API as it is the
+    simplest to implement for low-QPS services.
+
+*   **Favor callback API over other APIs for all RPCs**, given that the
+    application can avoid all blocking operations or blocking operations can be
+    moved to a separate thread.
+
+*   If having to use the async completion-queue API, the **best scalability
+    trade-off is having numcpu’s threads and one completion queue per thread**.
+
+*   For the async completion-queue API, make sure to **register enough server
+    requests for the desired level of concurrency** to avoid the server
+    continuously getting stuck in a slow path that results in essentially serial
+    request processing. 
+
+*   **Enable write batching in streams** if message k + 1 does not rely on
+    responses from message k by passing a WriteOptions argument to Write with
+    buffer_hint set - `stream_writer->Write(message,
+    WriteOptions().set_buffer_hint());`
+
+*   *(Special topic)* In the case that your application has unary or server-side
+    streaming methods where there will be exactly one request message from the
+    client, the **synchronous-with-deferred-read interface** could reduce memory
+    usage on busy servers. Read more about the interface
+    [here](https://g3doc.corp.google.com/net/grpc/g3doc/cpp/codegen.md#synchronous-with-deferred-read-interface).
```

Comment:
@jtattermusch I wasn't able to find a link for the sync-with-deferred-read interface in github, any idea what I could link here?

Reply:
I don't think there is currently anything in OSS about this.  But I'd really love to see our generated code guide moved to OSS, since it's incredibly useful (for a lot more than just this one topic).

Reply:
Maybe just omit the line for now and port the guide over from Google3 in a follow-on PR? I think we've never had an actual user for this interface that we know of. It was a feature request but the project that wanted it ended up going nowhere.

Reply:
I don't have bandwidth to move the code guide over at the moment, but I'll create bugs for this internally for when I can get to this. In the meantime, I'll omit this topic as it seems fairly low-priority.